internal/backend/remote-state/assure/auth/auth.go

```diff
@@ -0,0 +1,60 @@
+package auth
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/opentofu/opentofu/internal/tfdiags"
+)
+
+type Config struct {
+	*AzureCLIAuthConfig
+	*ClientBasicAuthConfig
+	*ClientCertificateAuthConfig
+	*OIDCAuthConfig
+	*MSIAuthConfig
+	*StorageAddresses
+}
+
+type AuthMethod interface {
+	// Construct takes the configuration and obtains an Azure-native
+	// authentication method, appropriate for the Azure sdk's various clients.
+	Construct(ctx context.Context, config *Config) (azcore.TokenCredential, error)
+
+	// Validate ensures this authentication method has the configuration variables and is
+	// the appropriate method to use. A nil return for diagnostics implies that there is no
+	// need to look further for authentication methods.
+	Validate(config *Config) tfdiags.Diagnostics
+
+	// AugmentConfig should be called to ensure the config has all proper storage names
+	// when attempting to get the storage account's access keys. It will return an error if
+	// the expected storage names, IDs, and addresses are not present.
+	//
+	// Note: only the CLI is really able to actually *change* the config, by obtaining information
+	// out of the azure profile saved on the filesystem.
+	AugmentConfig(config *Config) error
+}
+
+func GetAuthMethod(config *Config) (AuthMethod, error) {
+	authMethods := []AuthMethod{
+		&clientCertAuth{},
+		&clientBasicAuth{},
+		&oidcAuth{},
+		&managedIdentityAuth{},
+		&azureCLICredentialAuth{},
+	}
+	var diags tfdiags.Diagnostics
+	for _, authMethod := range authMethods {
+		if d := authMethod.Validate(config); d.HasErrors() {
+			diags = diags.Append(d)
+			continue
+		}
+		return authMethod, nil
+	}
+	diags = diags.Append(tfdiags.Sourceless(
```

Comment:
We should be able to get a general hcl.Range into here from *Config so it's clear it comes from the backend block

Reply:
An alternate approach is to add context to all diagnostics generated in this package with diags.InConfigBody(body, addr) external to all the azure specific logic

Reply:
I'm having trouble finding a method in `ResourceData` which obtains an `hcl.Range` or address, could you point me in the right direction? Diagnostics with the line number seems like a useful feature to have.